Modules/_csv.c

```diff
@@ -1342,75 +1281,98 @@ csv_writerow(PyObject *op, PyObject *seq)
+            str_field = Py_NewRef(field);
+        } else {
+            str_field = PyObject_Str(field);
+            if (str_field == NULL) {
+                Py_DECREF(field);
+                goto error;
+            }
         }
-        else if (null_field) {
-            append_ok = join_append(self, NULL, quoted);
-            Py_DECREF(field);
+
+        /* Single empty field special case */
+        if (field_count == 0) {
+            first_field_was_none = is_none;
+            if (is_none) {
+                first_field_was_empty_like = true;
+            } else {
+                Py_ssize_t len = str_field ? PyUnicode_GET_LENGTH(str_field) : 0;
+                first_field_was_empty_like = (len == 0);
+            }
         }
-        else {
-            PyObject *str;
 
-            str = PyObject_Str(field);
+        /* Important side-effect: this may promote to "quoted". */
+        if (_write_field(writer, self, str_field, &quoted) < 0) {
+            Py_XDECREF(str_field);
             Py_DECREF(field);
-            if (str == NULL) {
-                Py_DECREF(iter);
-                return NULL;
-            }
-            append_ok = join_append(self, str, quoted);
-            Py_DECREF(str);
+            goto error;
         }
-        if (!append_ok) {
-            Py_DECREF(iter);
-            return NULL;
+
+        if (field_count == 0) {
+            first_field_was_quoted_in_loop = (quoted != 0);
         }
+
+        Py_XDECREF(str_field);
+        Py_DECREF(field);
+        field_count++;
+    }
+
+    if (PyErr_Occurred()) {
+        goto error;
     }
     Py_DECREF(iter);
-    if (PyErr_Occurred())
-        return NULL;
+    iter = NULL;
 
-    if (self->num_fields > 0 && self->rec_len == 0) {
+    /* Single empty-field rule */
+    if (field_count == 1 && first_field_was_empty_like && !first_field_was_quoted_in_loop) {
         if (dialect->quoting == QUOTE_NONE ||
-            (null_field &&
+            (first_field_was_none &&
              (dialect->quoting == QUOTE_STRINGS ||
-              dialect->quoting == QUOTE_NOTNULL)))
-        {
-            PyErr_Format(self->error_obj,
-                "single empty field record must be quoted");
+              dialect->quoting == QUOTE_NOTNULL))) {
+            PyErr_SetString(self->error_obj,
+                            "single empty field record must be quoted");
+            PyUnicodeWriter_Discard(writer);
             return NULL;
         }
-        self->num_fields--;
-        if (!join_append(self, NULL, 1))
-            return NULL;
+        if (PyUnicodeWriter_WriteChar(writer, dialect->quotechar) < 0) goto error_after_iter;
+        if (PyUnicodeWriter_WriteChar(writer, dialect->quotechar) < 0) goto error_after_iter;
     }
 
-    /* Add line terminator.
-     */
-    if (!join_append_lineterminator(self)) {
+    if (PyUnicodeWriter_WriteStr(writer, self->dialect->lineterminator) < 0) {
+error_after_iter:
```

Comment:
I don't like a label inside a control flow, please avoid this.

Reply:
1836417e122ea5a5299937ee7f59687a7e9cff37